File: toolargetool/src/main/java/com/gu/toolargetool/Formatter.java
package com.gu.toolargetool;

import android.app.Activity;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.os.Bundle;

/**
 * Created by sfriedenberg on 09/20/2017.
 */

public abstract class Formatter {
    public abstract String format(Activity activity, Bundle bundle);
    public abstract String format(FragmentManager fragmentManager, Fragment fragment, Bundle bundle);

    public static Formatter DEFAULT_FORMATTER = new Formatter() {
        @Override
        public String format(Activity activity, Bundle bundle) {
            return activity.getClass().getSimpleName() + ".onSaveInstanceState wrote: " + TooLargeTool.bundleBreakdown(bundle);
        }

        @Override
        public String format(FragmentManager fragmentManager, Fragment fragment, Bundle bundle) {
            String message = fragment.getClass().getSimpleName() + ".onSaveInstanceState wrote: " + TooLargeTool.bundleBreakdown(bundle);
            Bundle fragmentArguments = fragment.getArguments();
            if (fragmentArguments != null) {
                message += "\n* fragment arguments = " + TooLargeTool.bundleBreakdown(fragmentArguments);
            }

            return message;
        }
    };
}

File: toolargetool/src/main/java/com/gu/toolargetool/Formatter.kt
package com.gu.toolargetool

import android.app.Activity
import android.os.Bundle
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager

/**
 * Interface that allows flexibility in how TooLargeTool's output is formatted. The default
 * implementation [DefaultFormatter] should be suitable in most cases.
 */
interface Formatter {
    fun format(activity: Activity, bundle: Bundle): String
    fun format(fragmentManager: androidx.fragment.app.FragmentManager, fragment: androidx.fragment.app.Fragment, bundle: Bundle): String
}

/**
 * The default implementation of [Formatter].
 *
 * @author [@sfriedenberg](https://github.com/friedenberg)
 */
class DefaultFormatter: Formatter {
    override fun format(activity: Activity, bundle: Bundle): String {
        return activity.javaClass.simpleName + ".onSaveInstanceState wrote: " + TooLargeTool.bundleBreakdown(bundle)
    }

    override fun format(fragmentManager: androidx.fragment.app.FragmentManager, fragment: androidx.fragment.app.Fragment, bundle: Bundle): String {
        var message = fragment.javaClass.simpleName + ".onSaveInstanceState wrote: " + TooLargeTool.bundleBreakdown(bundle)
        val fragmentArguments = fragment.arguments
        if (fragmentArguments != null) {
            message += "\n* fragment arguments = " + TooLargeTool.bundleBreakdown(fragmentArguments)
        }

        return message
    }
}
